prov/rstream/src/rstream_msg.c

```diff
@@ -484,38 +496,46 @@ static ssize_t rstream_send(struct fid_ep *ep_fid, const void *buf, size_t len,
 	char *remote_addr = NULL;
 	size_t sent_len = 0;
 	uint32_t curr_avail_len = len;
+	void *ctx;
 
 	do {
-		if (!rstream_can_send(ep)) {
-			ret = retry_send_resources(ep);
-			if (ret < 0) {
-				if(ret < 0 && ret != -FI_EAGAIN)
-					return ret;
-				else
-					return ((sent_len) ? sent_len : ret);
-			}
+		ret = rstream_can_send(ep);
```

Comment:
There's no lock held around this call.  So multiple threads could check it, all get okay, then proceed.  Is this safe?

Reply:
the locking is within the can_send call; I did it this way because I have to handle the CQ polling case. In this case I must release the tx resource lock so I don't deadlock with a control message or tx update. However, since another thread could grab the tx resource my cq polling thread found, I must recheck my resource status after coming out of can_send and after taking the lock before I can send anything.

Reply:
The intention is to check the CQ for an update if resources are exhausted.  The result can enable a send to initiate, but there is no guarantee. The way its currently structured doesn't force ordering between threaded send requests if resources become available, but it avoids a race on acquiring the resource. I'll work on restructuring this area.